### ISSUE TITLE
copilot svc override doesn't add resources on cloudfront distribution

### ISSUE THREAD

**Issue by acamb** [aws/copilot-cli#4971]
Hi,
I'm using copilot svc override to add a ssl certificate from ACM to the cloudfront distribution of a "Static Site" service.
I've used both cdk and yamlpatch as override tools with no results: the "copilot svc package --diff" command shows no changes.

CDK snippet:
```typescript
    transformCloudFrontDistribution() {
        const cloudFrontDistribution = this.template.getResource("CloudFrontDistribution") as cf.CfnDistribution;
        cloudFrontDistribution.addOverride("Properties.CloudFrontDistribution.DistributionConfig.Aliases","xxx.xxx.com");
        cloudFrontDistribution.addOverride("Properties.CloudFrontDistribution.DistributionConfig.ViewerCertificate.AcmCertificateArn","arn:aws:acm:us-east-1:xxxx");
    }
``` 

yamlpatch snippet:
```
- op: add
  path: /Resources/CloudFrontDistribution/Properties/DistributionConfig/Aliases
  value: "xxx.xxx.com"
- op: add
  path: /Resources/CloudFrontDistribution/Properties/DistributionConfig/ViewerCertificate/AcmCertificateArn
  value: "arn:aws:acm:us-east-1:xxxx"
``` 

I'm missing something?

Kind regards, 
Andrea

**Comment by bvtujo** [comment]
Hey @acamb, the CDK overrides should be specified starting from the `Properties` key of the raw CFN snippet, like so: 

```yaml
CloudFrontDistribution:
  Metadata:
    'aws:copilot:description': 'A CloudFront distribution for global content delivery'
  Type: AWS::CloudFront::Distribution
  Properties:
    DistributionConfig:
      Aliases: [{{.StaticSiteAlias}}]
      DefaultCacheBehavior:
        Compress: true
        AllowedMethods: ["GET", "HEAD"]
        FunctionAssociations:
          - EventType: viewer-request
            FunctionARN: !GetAtt CloudFrontViewerRequestRewriteFunction.FunctionARN
        ViewerProtocolPolicy: redirect-to-https
        CachePolicyId: 658327ea-f89d-4fab-a63d-7e88639e58f6 # See https://go.aws/3bJid3k
        TargetOriginId: !Sub 'copilot-${AppName}-${EnvName}-${WorkloadName}'
      ...
```

So, to override the `Aliases` property, you'd grab the `CloudFrontDistribution` resource as above, but then specify 
```
cloudFrontDistribution.addOverride("Properties.DistributionConfig.Aliases","xxx.xxx.com");
cloudFrontDistribution.addOverride("Properties.DistributionConfig.ViewerCertificate.AcmCertificateArn","arn:aws:acm:us-east-1:xxxx");
```

I'm not sure why Yamlpatch isn't working for you, since that syntax looks correct. Are you specifying an `alias` in the manifest? If you are, then the `alias` key will already exist and the `add` action may be a no-op. 

When you specify `alias` in the manifest, the `Aliases` and `ViewerCertificate` properties of the CFN resource are populated. If you don't specify an alias, they aren't. Could that be what's going on here? You can inspect the CFN template for yourself by running `copilot svc package` and looking for the `CloudFrontDistribution` resource in the template.

**Comment by RaffaeleSgarro** [comment]
@bvtujo did you actually **try it** with the latest version of Copilot or just speculating? Because if it's just speculation that habit explains a lot of the issues we are having with Copilot, which is full of things that don't work as advertised, and it would just require **actual** testing to discover them, whereas speculation or synthetic `_test.go` won't.

**Comment by acamb** [comment]
Hi @bvtujo,
I've updated my cdk override with your suggestions:
```typescript
    transformCloudFrontDistribution() {
        const cloudFrontDistribution = this.template.getResource("CloudFrontDistribution") as cf.CfnDistribution;
        cloudFrontDistribution.addOverride("Properties.DistributionConfig.Aliases","xxx.xxx.com");
        cloudFrontDistribution.addOverride("Properties.DistributionConfig.ViewerCertificate.AcmCertificateArn","arn:aws:acm:us-east-1:xxxx");
    }
``` 

I'm not specifying an alias in the manifest.

I can confirm that nor cdk neither yamlpatch are adding the override to the cloudformation template.
The template generated with `copilot svc package --name [...]` is:

```
...

Resources:
  
  ...

  CloudFrontOriginAccessControl:
    Metadata:
      'aws:copilot:description': 'Access control to make the content in the S3 bucket only accessible through CloudFront'
    Type: AWS::CloudFront::OriginAccessControl
    Properties:
      OriginAccessControlConfig:
        Description: !Sub 'Access control for static s3 origin for ${AppName}-${EnvName}-${WorkloadName}'
        # Truncate the name to allow at most 64 characters.
        Name: datasinc-ecs-prod-app-awards
        OriginAccessControlOriginType: s3
        SigningBehavior: always
        SigningProtocol: sigv4

  CloudFrontViewerRequestRewriteFunction:
    Metadata:
      'aws:copilot:description': 'CloudFront Function to rewrite viewer request to index.html'
    Type: AWS::CloudFront::Function
    Properties:
      AutoPublish: true
      FunctionCode: |
        function handler(event){var request=event.request;var uri=request.uri;if(uri.endsWith('/')){request.uri+='index.html'}else if(!uri.includes('.')){request.uri+='/index.html'}return request}
      FunctionConfig:
        Comment: CloudFront Function to rewrite viewer request to index.html
        Runtime: cloudfront-js-1.0
      # Truncate the name to allow at most 64 characters.
      Name: datasinc-ecs-prod-app-awards

  CloudFrontDistribution:
    Metadata:
      'aws:copilot:description': 'A CloudFront distribution for global content delivery'
    Type: AWS::CloudFront::Distribution
    Properties:
      DistributionConfig:
        DefaultCacheBehavior:
          Compress: true
          AllowedMethods: ["GET", "HEAD"]
          FunctionAssociations:
            - EventType: viewer-request
              FunctionARN: !GetAtt CloudFrontViewerRequestRewriteFunction.FunctionARN
          ViewerProtocolPolicy: redirect-to-https
          CachePolicyId: 658327ea-f89d-4fab-a63d-7e88639e58f6 # See https://go.aws/3bJid3k
          TargetOriginId: !Sub 'copilot-${AppName}-${EnvName}-${WorkloadName}'
        Enabled: true
        IPV6Enabled: true
        Origins:
          - Id: !Sub 'copilot-${AppName}-${EnvName}-${WorkloadName}'
            DomainName: !GetAtt Bucket.RegionalDomainName
            OriginAccessControlId: !Ref CloudFrontOriginAccessControl
            # Workaround for using Origin Access Control as Origin Access Identity is still
            # required when the origin is an S3 bucket.
            S3OriginConfig:
              OriginAccessIdentity: ''
     ...
Outputs:
  CloudFrontDistributionDomainName:
    Value: !GetAtt CloudFrontDistribution.DomainName
    Export:
      Name: !Sub ${AWS::StackName}-CloudFrontDistributionDomainName


```

**Comment by bvtujo** [comment]
@acamb Here's what worked for me in 1.28:

edit: apologies, I tested this with env overrides, not static site. Please hold on while I retry. 
edit2: It does look like the CDK app isn't getting run when I do `copilot svc package` on the static site pattern. 

#### CDK Override at `copilot/environments/overrides/stack.ts`

```typescript
transformCloudFrontDistribution() {
    const cloudFrontDistribution = this.template.getResource("CloudFrontDistribution") as cf.CfnDistribution;
    cloudFrontDistribution.addOverride("Properties.DistributionConfig.Aliases","xxx.xxx.com");
    cloudFrontDistribution.addOverride("Properties.DistributionConfig.ViewerCertificate.AcmCertificateArn","arn:aws:acm:us-east-1:xxxx");
}
```

<details>
<summary>package.json</summary>

```json
{
  "name": "override",
  "version": "0.1.0",
  "bin": {
    "override": "bin/override.js"
  },
  "scripts": {
    "build": "tsc",
    "watch": "tsc -w",
    "cdk": "cdk"
  },
  "devDependencies": {
    "@types/jest": "^29.2.4",
    "@types/node": "18.11.15",
    "jest": "^29.3.1",
    "ts-jest": "^29.0.3",
    "aws-cdk": "2.56.0",
    "ts-node": "^10.9.1",
    "typescript": "~4.9.4"
  },
  "dependencies": {
    "aws-cdk-lib": "2.56.0",
    "constructs": "^10.0.0",
    "source-map-support": "^0.5.21"
  }
}
```

</details>

<details>
<summary>Full typescript file</summary>

```typescript
import * as cdk from 'aws-cdk-lib';
import * as path from 'path';
import { aws_cloudfront as cf } from 'aws-cdk-lib';

interface TransformedStackProps extends cdk.StackProps {
    readonly appName: string;
    readonly envName: string;
}

export class TransformedStack extends cdk.Stack {
    public readonly template: cdk.cloudformation_include.CfnInclude;
    public readonly appName: string;
    public readonly envName: string;

    constructor (scope: cdk.App, id: string, props: TransformedStackProps) {
        super(scope, id, props);
        this.template = new cdk.cloudformation_include.CfnInclude(this, 'Template', {
            templateFile: path.join('.build', 'in.yml'),
        });
        this.appName = props.appName;
        this.envName = props.envName;
        this.transformCloudFrontDistribution();
    }
    
    // TODO: implement me.
    transformCloudFrontDistribution() {
        const cloudFrontDistribution = this.template.getResource("CloudFrontDistribution") as cf.CfnDistribution;
        cloudFrontDistribution.addOverride("Properties.DistributionConfig.Aliases","xxx.xxx.com");
        cloudFrontDistribution.addOverride("Properties.DistributionConfig.ViewerCertificate.AcmCertificateArn","arn:aws:acm:us-east-1:xxxx");
    }    
}
```

</details>

#### And the resulting CDN snippets in the template
```
DistributionConfig:
  Aliases: xxx.xxx.com
  ...
  ViewerCertificate:
    AcmCertificateArn: arn:aws:acm:us-east-1:xxxx
```
<details>
<summary>Full resource description</summary>

```yaml
CloudFrontDistribution:
  Type: AWS::CloudFront::Distribution
  Properties:
    DistributionConfig:
      Aliases: xxx.xxx.com
      DefaultCacheBehavior:
        AllowedMethods:
          - GET
          - HEAD
          - OPTIONS
          - PUT
          - PATCH
          - POST
          - DELETE
        CachePolicyId: 4135ea2d-6df8-44a3-9df3-4b5a84be39ad
        OriginRequestPolicyId: 216adef6-5c7f-47e4-b989-5492eafa07d3
        TargetOriginId:
          Fn::Sub: copilot-${AppName}-${EnvironmentName}-origin
        ViewerProtocolPolicy: allow-all
      Enabled: true
      IPV6Enabled: true
      Origins:
        - CustomOriginConfig:
            OriginProtocolPolicy: match-viewer
          DomainName:
            Fn::GetAtt: PublicLoadBalancer.DNSName
          Id:
            Fn::Sub: copilot-${AppName}-${EnvironmentName}-origin
      ViewerCertificate:
        AcmCertificateArn: arn:aws:acm:us-east-1:xxxx
```
</details>

**Comment by bvtujo** [comment]
@acamb it looks like we have a bug in our code which doesn't override static sites during `svc package`. The logic does run during `copilot svc deploy`, however. 

We've merged a fix for this (#4952) which will go out in our next release.

**Comment by bvtujo** [comment]
Closing as resolved due to the fix being merged; please feel free to reopen if there are other problems you notice with static site overrides. You can also view the template diff with the `copilot svc deploy --diff` flag, which should show you any changes due to overrides before the deployment proceeds.
